### DISCUSSION TITLE
How to use UserDataQos on EDP layer

### DISCUSSION THREAD

**Discussion by advancenXX** [eProsima/Fast-DDS#3258] [Q&A]
We could add UserData to DomainParticipantQos when create_participant(), and receive it on_participant_discovery().
We could also add UserData to DataWriterQos when create_datawriter(), but how could we receive it?

**Comment by JLBuenoLopez** [comment]
Hi @advancenXX 

`DomainParticipantListener` class implements also the `on_subscriber_discovery` and `on_publisher_discovery` callbacks. I know that these names might be misleading because DDS specification uses the terms Subscriber and Publisher as the factories for the DataWriter and DataReader endpoints. These names are legacy from the old FastRTPS API (non-DDS compliant). These callbacks are called when a DDS DataWriter or DDS DataReader is discovered (and not when a DDS Publisher/Subscriber is). You might access the UserData QoS information in a similar way as in the `on_participant_discovery` callback.

I hope this information answers your question.

**Reply by advancenXX** [reply]
Thanks, I find on_publisher_discovery too and that is what I want.

**Comment by advancenXX** [comment]
@JLBuenoLopez-eProsima hello, I add a on_publisher_discovery participant listener simply to HelloWorldExample, both for Publisher and Subscriber side, like,
DomainParticipantFactory::get_instance()->create_participant(0, pqos, &part_listener_);
Then, listener works correctly, but Subscriber could never receive data.
What could I do to fix this problem?

**Reply by EduPonz** [reply]
Hi @advancenXX,

This is because the default value for the `StatusMask` on participant creation is `all`, meaning the participant listener will capture all the callbacks coming from all its contained entities. Since your participant listener is not overriding the `on_data_on_readers` function, the default implementation it's being used, which is empty. You can read more about it [here](https://fast-dds.docs.eprosima.com/en/latest/fastdds/dds_layer/core/entity/entity.html#listener).

The easiest fix is to set the `StatusMask` to `none` when creating your participant, as the `on_participant|publisher|subscriber_discovery` callback are not affected by it.

**Reply by advancenXX** [reply]
Thanks, I will try it on my device.
